src/extension/index.ts

```diff
@@ -0,0 +1,239 @@
+import type { AppInfo } from "../appInfo.js"
+import type { CID } from "../common/cid.js"
+import type { DistinctivePath, Partition } from "../path/index.js"
+import type { Maybe } from "../common/types.js"
+import type { Permissions } from "../permissions.js"
+import type { Session } from "../session.js"
+import type { ShortHands } from "../index.js"
+
+import * as Events from "../events.js"
+import { VERSION } from "../index.js"
+
+
+// CREATE
+
+
+type Config = {
+  namespace: AppInfo | string
+  session: Maybe<Session>
+  capabilities?: Permissions
+  shorthands: ShortHands
+  lookupDataRoot: (username: string) => Promise<CID | null>
+  eventEmitters: {
+    fileSystem: Events.Emitter<Events.FileSystem>
+    session: Events.Emitter<Events.Session<Session>>
+  }
+}
+
+export async function create(config: Config): Promise<{
+  connect: (extensionId: string) => void
+  disconnect: (extensionId: string) => void
+}> {
+  let connection: Connection = { extensionId: null, connected: false }
+  let listeners: Listeners
+
+  return {
+    connect: async (extensionId: string) => {
+      connection = await connect(extensionId, config)
+
+      // The extension may call connect more than once, but
+      // listeners should only be added once
+      if (!listeners) listeners = listen(connection, config)
+    },
+    disconnect: async (extensionId: string) => {
+      connection = await disconnect(extensionId, config)
+      stopListening(config, listeners)
+    }
+  }
+}
+
+
+
+// CONNECTION
+
+
+type Connection = {
+  extensionId: string | null
+  connected: boolean
+}
+
+async function connect(extensionId: string, config: Config): Promise<Connection> {
+  const state = await getState(config)
+
+  globalThis.postMessage({
+    id: extensionId,
+    type: "connect",
+    timestamp: Date.now(),
+    state
+  })
+
+  return { extensionId, connected: true }
+}
+
+async function disconnect(extensionId: string, config: Config): Promise<Connection> {
+  const state = await getState(config)
+
+  globalThis.postMessage({
+    id: extensionId,
+    type: "disconnect",
+    timestamp: Date.now(),
+    state
+  })
+
+  return { extensionId, connected: false }
+}
+
+
+
+// LISTENERS
+
+
+type Listeners = {
+  handleLocalChange: (params: { root: CID; path: DistinctivePath<[ Partition, ...string[] ]> }) => Promise<void>
+  handlePublish: (params: { root: CID }) => Promise<void>
+  handleSessionCreate: (params: { session: Session }) => Promise<void>
+  handleSessionDestroy: (params: { username: string }) => Promise<void>
+}
+
+function listen(connection: Connection, config: Config): Listeners {
+  async function handleLocalChange(params: { root: CID; path: DistinctivePath<[ Partition, ...string[] ]> }) {
+    const { root, path } = params
+    const state = await getState(config)
+
+    globalThis.postMessage({
+      id: connection.extensionId,
+      type: "filesystem",
```

Comment:
Maybe `fileSystem` with a capital S to be consistent?

Reply:
Yep, good call. Updated that. 👌